File: server/src/services/RoomManager.ts
import WebSocket from "ws";
import { customAlphabet, nanoid } from "nanoid";
import crypto from "crypto";
import { CustomWebSocket, ImageChatMessage, ParticipantRecord, ServerPayload, TextChatMessage } from "../types/room.js";
import { Room } from "../models/Room.js";
import { messageStore } from "./messageService.js";

const RECONNECT_GRACE_PERIOD_MS = 30 * 1000;
const MAX_PARTICIPANTS_PER_ROOM = 50;
const MAX_ROOMS = 1000;
const RATE_LIMIT_WINDOW_MS = 10_000;
const RATE_LIMIT_MAX_MESSAGES = 20;
const ROOM_ID_RESERVATION_MS = 30 * 1000;
const ROOM_ID_GENERATION_ATTEMPTS = 5;
const defaultCreateRoomId = customAlphabet("abcdefghijklmnopqrstuvwxyz0123456789-", 10);
const USERNAME_REGEX = /^[a-zA-Z0-9_]{5,25}$/;
const ROOM_REGEX = /^[a-zA-Z0-9_-]{5,35}$/;

interface JoinResult {
    roomId: string;
    participantId: string;
    reconnectToken: string;
    username: string;
    userCount: number;
    participants: ReturnType<Room["getParticipantView"]>;
    history: TextChatMessage[];
    reconnected: boolean;
    historyEnabled: boolean;
}

class RoomManager {
    public rooms: Map<string, Room> = new Map();
    private pendingRoomReservations: Map<string, ReturnType<typeof setTimeout>> = new Map();
    private roomIdGenerator: () => string = defaultCreateRoomId;

    public createRoomId(): string {
        if (this.rooms.size + this.pendingRoomReservations.size >= MAX_ROOMS) {
            throw new Error("server at capacity; try again later");
        }

        for (let attempt = 0; attempt < ROOM_ID_GENERATION_ATTEMPTS; attempt += 1) {
            const roomId = this.roomIdGenerator();
            if (this.rooms.has(roomId) || this.pendingRoomReservations.has(roomId)) {
                continue;
            }

            const timer = setTimeout(() => {
                this.pendingRoomReservations.delete(roomId);
            }, ROOM_ID_RESERVATION_MS);

            this.pendingRoomReservations.set(roomId, timer);
            return roomId;
        }

        throw new Error("failed to allocate room id; try again");
    }

    private async getOrCreateRoom(roomId: string): Promise<Room> {
        let room = this.rooms.get(roomId);
        if (!room) {
            const hasReservation = this.pendingRoomReservations.has(roomId);
            if (this.rooms.size + this.pendingRoomReservations.size >= MAX_ROOMS && !hasReservation) {
                throw new Error("server at capacity; try again later");
            }

            this.releasePendingRoomReservation(roomId);
            room = new Room(roomId, Date.now());
            this.rooms.set(roomId, room);
            await messageStore.deleteRoomMessages(roomId);
        }
        return room;
    }

    public async joinRoom(
        roomId: string,
        username: string,
        ws: WebSocket,
        participantId?: string,
        reconnectToken?: string
    ): Promise<JoinResult> {
        if (!ROOM_REGEX.test(roomId)) {
            throw new Error("invalid room id");
        }

        if (!USERNAME_REGEX.test(username)) {
            throw new Error("invalid username");
        }

        const socket = ws as CustomWebSocket;
        const room = await this.getOrCreateRoom(roomId);

        const reconnect = this.getReconnectParticipant(room, username, participantId, reconnectToken);
        const active = this.getActiveParticipant(room, username, participantId, reconnectToken);

        if (room.userCount >= MAX_PARTICIPANTS_PER_ROOM && !reconnect && !active) {
            throw new Error("room is full");
        }

        if (!reconnect && !active && room.hasUsername(username)) {
            throw new Error("username already taken");
        }

        let participant: ParticipantRecord;
        let token = reconnectToken || "";
        let reconnected = false;

        if (reconnect) {
            clearTimeout(reconnect.timer);
            room.disconnectedUsers.delete(reconnect.participant.participantId);
            participant = reconnect.participant;
            reconnected = true;
        } else if (active) {
            active.socket!.replaced = true;
            active.socket!.close(1000, "Session replaced");
            participant = active;
            reconnected = true;
        } else {
            token = nanoid(32);
            participant = {
                participantId: nanoid(16),
                username,
                reconnectTokenHash: this.hashReconnectToken(token),
                connected: true,
                lastSeenAt: Date.now(),
                socket: null,
                messageTimestamps: []
            };
        }

        room.addParticipant(participant, socket);
        const history = await messageStore.getMessages(roomId);

        const joinResult = {
            roomId,
            participantId: participant.participantId,
            reconnectToken: token,
            username: participant.username,
            userCount: room.userCount,
            participants: room.getParticipantView(),
            history,
            reconnected,
            historyEnabled: messageStore.isEnabled
        };

        return joinResult;
    }

    public finalizeJoin(roomId: string, username: string, reconnected: boolean): void {
        const room = this.rooms.get(roomId);
        if (!room) return;

        this.broadcastParticipants(room);
        this.broadcastSystem(room, `${username} ${reconnected ? "reconnected" : "joined"}`);
    }

    public async leaveRoom(roomId: string, ws: WebSocket): Promise<void> {
        const socket = ws as CustomWebSocket;
        const room = this.rooms.get(roomId);
        if (!room || !socket.participantId) return;

        const participant = room.removeParticipant(socket.participantId);
        socket.roomId = null;
        socket.participantId = null;
        socket.username = null;

        if (participant) {
            this.broadcastSystem(room, `${participant.username} left`);
            this.broadcastParticipants(room);
        }

        await this.deleteRoomIfEmpty(roomId, room);
    }

    public markDisconnected(roomId: string, ws: WebSocket): void {
        const socket = ws as CustomWebSocket;
        if (socket.replaced) return;

        const room = this.rooms.get(roomId);
        if (!room || !socket.participantId) return;

        if (socket.intentionalLeave) {
            void this.leaveRoom(roomId, socket);
            return;
        }

        const participant = room.removeParticipant(socket.participantId);
        if (!participant) return;

        const timer = setTimeout(() => {
            room.disconnectedUsers.delete(participant.participantId);
            this.broadcastParticipants(room);
            void this.deleteRoomIfEmpty(roomId, room);
        }, RECONNECT_GRACE_PERIOD_MS);

        room.disconnectedUsers.set(participant.participantId, {
            participant,
            timer
        });

        this.broadcastSystem(room, `${participant.username} disconnected`);
        this.broadcastParticipants(room);
    }

    public async createMessage(roomId: string, ws: WebSocket, text: string): Promise<TextChatMessage> {
        const socket = ws as CustomWebSocket;
        const room = this.rooms.get(roomId);
        if (!room || !socket.participantId) {
            throw new Error("room not joined");
        }

        const participant = room.participants.get(socket.participantId);
        if (!participant) {
            throw new Error("participant not joined");
        }

        const message: TextChatMessage = {
            id: crypto.randomUUID(),
            type: "MESSAGE_CREATED",
            participantId: participant.participantId,
            username: participant.username,
            contentType: "text",
            text,
            timestamp: Date.now()
        };

        room.broadcast(message);
        void messageStore.saveMessage(roomId, message);
        return message;
    }

    public createImageMessage(
        roomId: string,
        participantId: string,
        image: ImageChatMessage["image"]
    ): ImageChatMessage {
        const room = this.rooms.get(roomId);
        if (!room) {
            throw new Error("room not joined");
        }

        const participant = room.participants.get(participantId);
        if (!participant) {
            throw new Error("participant not joined");
        }

        const message: ImageChatMessage = {
            id: crypto.randomUUID(),
            type: "MESSAGE_CREATED",
            participantId: participant.participantId,
            username: participant.username,
            contentType: "image",
            image,
            timestamp: Date.now()
        };

        room.broadcast(message);
        return message;
    }

    public sendToUser(ws: WebSocket, payload: ServerPayload): void {
        if (ws.readyState === WebSocket.OPEN) {
            ws.send(JSON.stringify(payload));
        }
    }

    public isRateLimited(roomId: string, participantId: string): boolean {
        const room = this.rooms.get(roomId);
        const participant = room?.participants.get(participantId);
        if (!participant) {
            return false;
        }

        const now = Date.now();
        const recent = participant.messageTimestamps.filter(
            (timestamp) => now - timestamp < RATE_LIMIT_WINDOW_MS
        );

        if (recent.length >= RATE_LIMIT_MAX_MESSAGES) {
            participant.messageTimestamps = recent;
            return true;
        }

        recent.push(now);
        participant.messageTimestamps = recent;
        return false;
    }

    public getActiveParticipantForRequest(
        roomId: string,
        participantId: string,
        reconnectToken: string
    ): ParticipantRecord | null {
        const room = this.rooms.get(roomId);
        const participant = room?.participants.get(participantId);
        if (!participant || !this.isValidReconnectToken(participant, reconnectToken)) {
            return null;
        }

        return participant;
    }

    public clearPendingRoomReservations(): void {
        for (const timer of this.pendingRoomReservations.values()) {
            clearTimeout(timer);
        }
        this.pendingRoomReservations.clear();
    }

    public setRoomIdGeneratorForTesting(generator?: () => string): void {
        this.roomIdGenerator = generator ?? defaultCreateRoomId;
    }

    private getReconnectParticipant(
        room: Room,
        username: string,
        participantId?: string,
        reconnectToken?: string
    ) {
        if (!participantId || !reconnectToken) return null;
        const disconnected = room.disconnectedUsers.get(participantId);
        if (!disconnected) return null;
        if (disconnected.participant.username !== username) return null;
        if (!this.isValidReconnectToken(disconnected.participant, reconnectToken)) return null;
        return disconnected;
    }

    private getActiveParticipant(
        room: Room,
        username: string,
        participantId?: string,
        reconnectToken?: string
    ): ParticipantRecord | null {
        if (!participantId || !reconnectToken) return null;
        const participant = room.participants.get(participantId);
        if (!participant || participant.username !== username) return null;
        if (!this.isValidReconnectToken(participant, reconnectToken)) return null;
        return participant;
    }

    private isValidReconnectToken(participant: ParticipantRecord, reconnectToken: string): boolean {
        return crypto.timingSafeEqual(
            Buffer.from(participant.reconnectTokenHash),
            Buffer.from(this.hashReconnectToken(reconnectToken))
        );
    }

    private hashReconnectToken(token: string): string {
        return crypto.createHash("sha256").update(token).digest("hex");
    }

    private releasePendingRoomReservation(roomId: string): void {
        const timer = this.pendingRoomReservations.get(roomId);
        if (timer) {
            clearTimeout(timer);
            this.pendingRoomReservations.delete(roomId);
        }
    }

    private broadcastParticipants(room: Room): void {
        room.broadcast({
            type: "PARTICIPANTS_UPDATED",
            participants: room.getParticipantView(),
            userCount: room.userCount
        });
    }

    private broadcastSystem(room: Room, text: string): void {
        room.broadcast({
            type: "SYSTEM_NOTICE",
            text,
            timestamp: Date.now()
        });
    }

    private async deleteRoomIfEmpty(roomId: string, room: Room): Promise<void> {
        if (!room.isEmpty) return;

        room.clearAllReconnectTimers();
        this.rooms.delete(roomId);
        await messageStore.deleteRoomMessages(roomId);
    }
}

export const roomManager = new RoomManager();

File: backend-go/room_manager.go
package main

import (
	"fmt"
	"regexp"
	"strings"
	"sync"
	"time"
)

const (
	reconnectGracePeriod   = 30 * time.Second
	messageLimit           = 1000
	rateLimitWindowMS      = int64(10_000)
	rateLimitMaxMessages   = 20
	maxWSPayloadBytes      = 16 * 1024
	maxParticipantsPerRoom = 50
	maxRooms               = 1000
)

var (
	usernamePattern = regexp.MustCompile(`^[a-zA-Z0-9_]{5,25}$`)
	roomPattern     = regexp.MustCompile(`^[a-zA-Z0-9_-]{5,35}$`)
)

type RoomManager struct {
	mu           sync.RWMutex
	rooms        map[string]*Room
	messageStore *MessageStore
}

type JoinResult struct {
	RoomID         string
	ParticipantID  string
	ReconnectToken string
	Username       string
	UserCount      int
	Participants   []ParticipantView
	History        []ChatMessage
	Reconnected    bool
	HistoryEnabled bool
}

func NewRoomManager(messageStore *MessageStore) *RoomManager {
	return &RoomManager{
		rooms:        make(map[string]*Room),
		messageStore: messageStore,
	}
}

func (m *RoomManager) CreateRoomID() string {
	return newRoomID()
}

func (m *RoomManager) getOrCreateRoom(roomID string) (*Room, error) {
	m.mu.Lock()
	defer m.mu.Unlock()

	room, ok := m.rooms[roomID]
	if ok {
		return room, nil
	}

	if len(m.rooms) >= maxRooms {
		return nil, fmt.Errorf("server at capacity; try again later")
	}

	room = NewRoom(roomID, time.Now().UnixMilli())
	m.rooms[roomID] = room
	m.messageStore.DeleteRoomMessages(roomID)
	return room, nil
}

func (m *RoomManager) JoinRoom(roomID, username string, socket *WSConn, participantID, reconnectToken string) (*JoinResult, error) {
	if !roomPattern.MatchString(roomID) {
		return nil, fmt.Errorf("invalid room id")
	}
	if !usernamePattern.MatchString(username) {
		return nil, fmt.Errorf("invalid username")
	}

	room, err := m.getOrCreateRoom(roomID)
	if err != nil {
		return nil, err
	}

	var participant *ParticipantRecord
	token := reconnectToken
	reconnected := false
	var previousSocket *WSConn

	room.mu.Lock()
	reconnect := m.getReconnectParticipantLocked(room, username, participantID, reconnectToken)
	active := m.getActiveParticipantLocked(room, username, participantID, reconnectToken)

	if reconnect == nil && active == nil && room.hasUsername(username) {
		room.mu.Unlock()
		return nil, fmt.Errorf("username already taken")
	}

	if room.userCount() >= maxParticipantsPerRoom && reconnect == nil && active == nil {
		room.mu.Unlock()
		return nil, fmt.Errorf("room is full")
	}

	if reconnect != nil {
		if reconnect.Timer != nil {
			reconnect.Timer.Stop()
		}
		delete(room.disconnectedUsers, reconnect.Participant.ParticipantID)
		participant = reconnect.Participant
		reconnected = true
	} else if active != nil {
		previousSocket = active.Socket
		participant = active
		reconnected = true
	} else {
		token = newReconnectToken()
		participant = &ParticipantRecord{
			ParticipantID:      newParticipantID(),
			Username:           username,
			ReconnectTokenHash: hashReconnectToken(token),
		}
	}

	room.addParticipant(participant, socket)
	result := &JoinResult{
		RoomID:         roomID,
		ParticipantID:  participant.ParticipantID,
		ReconnectToken: token,
		Username:       participant.Username,
		UserCount:      room.userCount(),
		Participants:   room.getParticipantView(),
		Reconnected:    reconnected,
		HistoryEnabled: m.messageStore.IsEnabled(),
	}
	room.mu.Unlock()

	if previousSocket != nil {
		previousSocket.MarkReplaced()
		_ = previousSocket.Close()
	}

	result.History = m.messageStore.GetMessages(roomID, historyMessageLimit)

	m.broadcastParticipants(room)
	var action string
	if reconnected {
		action = "reconnected"
	} else {
		action = "joined"
	}
	m.broadcastSystem(room, fmt.Sprintf("%s %s", participant.Username, action))
	return result, nil
}

func (m *RoomManager) LeaveRoom(roomID string, socket *WSConn) {
	room := m.getRoom(roomID)
	if room == nil {
		return
	}

	_, participantID, _, _, _ := socket.Metadata()
	if participantID == "" {
		return
	}

	var username string
	room.mu.Lock()
	participant := room.removeParticipant(participantID)
	socket.ClearParticipant()
	if participant != nil {
		username = participant.Username
	}
	room.mu.Unlock()

	if participant != nil {
		m.broadcastSystem(room, fmt.Sprintf("%s left", username))
		m.broadcastParticipants(room)
	}

	m.deleteRoomIfEmpty(roomID, room)
}

func (m *RoomManager) MarkDisconnected(roomID string, socket *WSConn) {
	room := m.getRoom(roomID)
	if room == nil {
		return
	}

	_, participantID, _, intentionalLeave, replaced := socket.Metadata()
	if replaced || participantID == "" {
		return
	}
	if intentionalLeave {
		m.LeaveRoom(roomID, socket)
		return
	}

	var participant *ParticipantRecord
	room.mu.Lock()
	participant = room.removeParticipant(participantID)
	if participant != nil {
		p := participant
		timer := time.AfterFunc(reconnectGracePeriod, func() {
			room.mu.Lock()
			delete(room.disconnectedUsers, p.ParticipantID)
			room.mu.Unlock()
			m.broadcastParticipants(room)
			m.deleteRoomIfEmpty(roomID, room)
		})
		room.disconnectedUsers[p.ParticipantID] = &DisconnectedUserInfo{
			Participant: p,
			Timer:       timer,
		}
	}
	room.mu.Unlock()

	if participant == nil {
		return
	}

	m.broadcastSystem(room, fmt.Sprintf("%s disconnected", participant.Username))
	m.broadcastParticipants(room)
}

func (m *RoomManager) CreateMessage(roomID string, socket *WSConn, text string) (*ChatMessage, error) {
	if len(text) == 0 {
		return nil, nil
	}
	if len(text) > messageLimit {
		return nil, fmt.Errorf("message too long (max %d characters)", messageLimit)
	}

	room := m.getRoom(roomID)
	if room == nil {
		return nil, fmt.Errorf("room not joined")
	}

	_, participantID, _, _, _ := socket.Metadata()
	if participantID == "" {
		return nil, fmt.Errorf("participant not joined")
	}

	room.mu.RLock()
	participant := room.participants[participantID]
	room.mu.RUnlock()
	if participant == nil {
		return nil, fmt.Errorf("participant not joined")
	}

	message := &ChatMessage{
		ID:            newUUIDv4(),
		Type:          "MESSAGE_CREATED",
		ParticipantID: participant.ParticipantID,
		Username:      participant.Username,
		Text:          text,
		Timestamp:     time.Now().UnixMilli(),
	}

	m.broadcastJSON(room, message)
	go m.messageStore.SaveMessage(roomID, *message)
	return message, nil
}

func (m *RoomManager) ValidateAndHandleMessage(socket *WSConn, payload sendMessage) *ErrorPayload {
	roomID, participantID, username, _, _ := socket.Metadata()
	if roomID == "" || participantID == "" || username == "" {
		return &ErrorPayload{Type: "ERROR", Code: "not_joined", Message: "join a room before sending messages"}
	}

	if rateLimited(socket) {
		return &ErrorPayload{Type: "ERROR", Code: "rate_limited", Message: "too many messages; slow down"}
	}

	text := strings.TrimSpace(payload.Text)
	if text == "" {
		return nil
	}
	if len(text) > messageLimit {
		return &ErrorPayload{Type: "ERROR", Code: "message_too_long", Message: fmt.Sprintf("message too long (max %d characters)", messageLimit)}
	}

	if _, err := m.CreateMessage(roomID, socket, text); err != nil {
		return &ErrorPayload{Type: "ERROR", Code: "handler_failed", Message: err.Error()}
	}
	return nil
}

func (m *RoomManager) getRoom(roomID string) *Room {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.rooms[roomID]
}

func (m *RoomManager) deleteRoomIfEmpty(roomID string, room *Room) {
	room.mu.RLock()
	empty := room.isEmpty()
	room.mu.RUnlock()
	if !empty {
		return
	}

	room.mu.Lock()
	room.clearAllReconnectTimers()
	room.mu.Unlock()

	m.mu.Lock()
	delete(m.rooms, roomID)
	m.mu.Unlock()
	m.messageStore.DeleteRoomMessages(roomID)
}

func (m *RoomManager) getReconnectParticipantLocked(room *Room, username, participantID, reconnectToken string) *DisconnectedUserInfo {
	if participantID == "" || reconnectToken == "" {
		return nil
	}
	disconnected := room.disconnectedUsers[participantID]
	if disconnected == nil {
		return nil
	}
	if disconnected.Participant.Username != username {
		return nil
	}
	if !validReconnectToken(disconnected.Participant, reconnectToken) {
		return nil
	}
	return disconnected
}

func (m *RoomManager) getActiveParticipantLocked(room *Room, username, participantID, reconnectToken string) *ParticipantRecord {
	if participantID == "" || reconnectToken == "" {
		return nil
	}
	participant := room.participants[participantID]
	if participant == nil || participant.Username != username {
		return nil
	}
	if !validReconnectToken(participant, reconnectToken) {
		return nil
	}
	return participant
}

func (m *RoomManager) broadcastParticipants(room *Room) {
	room.mu.RLock()
	payload := ParticipantsUpdatedPayload{
		Type:         "PARTICIPANTS_UPDATED",
		Participants: room.getParticipantView(),
		UserCount:    room.userCount(),
	}
	sockets := activeSocketsLocked(room)
	room.mu.RUnlock()
	sendToSockets(sockets, payload)
}

func (m *RoomManager) broadcastSystem(room *Room, text string) {
	room.mu.RLock()
	payload := SystemNoticePayload{
		Type:      "SYSTEM_NOTICE",
		Text:      text,
		Timestamp: time.Now().UnixMilli(),
	}
	sockets := activeSocketsLocked(room)
	room.mu.RUnlock()
	sendToSockets(sockets, payload)
}

func (m *RoomManager) broadcastJSON(room *Room, payload any) {
	room.mu.RLock()
	sockets := activeSocketsLocked(room)
	room.mu.RUnlock()
	sendToSockets(sockets, payload)
}

func activeSocketsLocked(room *Room) []*WSConn {
	sockets := make([]*WSConn, 0, len(room.participants))
	for _, participant := range room.participants {
		if participant.Socket != nil {
			sockets = append(sockets, participant.Socket)
		}
	}
	return sockets
}

func sendToSockets(sockets []*WSConn, payload any) {
	for _, socket := range sockets {
		_ = socket.SendJSON(payload)
	}
}

func rateLimited(socket *WSConn) bool {
	now := time.Now().UnixMilli()
	window := socket.RateLimitWindow()
	recent := make([]int64, 0, len(window)+1)
	for _, ts := range window {
		if now-ts < rateLimitWindowMS {
			recent = append(recent, ts)
		}
	}
	if len(recent) >= rateLimitMaxMessages {
		socket.SetRateLimitWindow(recent)
		return true
	}
	recent = append(recent, now)
	socket.SetRateLimitWindow(recent)
	return false
}
